File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
    Node head = new Node(0, 0), tail = new Node(0, 0);
    HashMap<Integer, Node> mpp = new HashMap<>();
    int capacity;
    
    private void add(Node cur) {
        mpp.put(cur.key, cur);
        Node headNext = head.next;
        head.next = cur;
        cur.prev = head;
        headNext.prev = cur;
        cur.next = headNext;
    }
    
    private void remove(Node cur) {
        mpp.remove(cur.key);
        cur.prev.next = cur.next;
        cur.next.prev = cur.prev;
    }
    
    class Node {
        Node prev, next;
        int key, value;
        Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    public LRUCache(int capacity) {
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        if (mpp.containsKey(key)) {
            Node cur = mpp.get(key);
            remove(cur);
            add(cur);
            return cur.value;
        }
        return -1;
    }
    
    public void put(int key, int value) {
        if (mpp.containsKey(key)) {
            remove(mpp.get(key));
        }
        if (mpp.size() == capacity) {
            remove(tail.prev);
        }
        add(new Node(key, value));
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.cpp
class Node {
public:
    int key;
    int val;
    Node* prev;
    Node* next;
    
    Node(int key, int val) {
        this->key = key;
        this->val = val;
        this->prev = nullptr;
        this->next = nullptr;
    }
};

class LRUCache {
public:
    unordered_map<int, Node*> xt;
    Node *head, *tail;
    int capacity;
    
    void add(Node* node) {
        xt[node->key] = node;
        Node* head_next = head->next;
        head->next = node;
        node->prev = head;
        node->next = head_next;
        head_next->prev = node;
    }
    
    void remove(Node* node) {
        xt.erase(node->key);
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }
    
    LRUCache(int capacity) {
        this->capacity = capacity;
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if (xt.find(key) == xt.end()) {
            return -1;
        }
        Node* cur = xt[key];
        remove(cur);
        add(cur);
        return cur->val;
    }
    
    void put(int key, int value) {
        if (xt.find(key) != xt.end()) {
            remove(xt[key]);
        } else if (xt.size() >= capacity) {
            remove(tail->prev);
        }
        Node* cur = new Node(key, value);
        add(cur);
    }
};